app/services/site_setting_update_existing_users.rb

```diff
@@ -164,4 +100,147 @@ def self.tag_notification_level(id)
+        .find_in_batches(batch_size: batch_size) do |users|
+          category_users =
+            users.map do |user|
+              { category_id: category_id, user_id: user.id, notification_level: notification_level }
+            end
+
+          CategoryUser.insert_all!(category_users)
+
+          processed_total += users.size
+          publish_progress(id, processed_total, total_users_to_process, modified_categories)
+        end
+    end
+
+    publish_progress(id, processed_total, total_users_to_process, modified_categories)
+  end
+
+  def self.default_tags(id, value, previous_value)
+    new_value = value.nil? ? "" : value
+
+    batch_size = 50_000
+
+    previous_tag_ids = Tag.where(name: previous_value.split("|")).pluck(:id)
+    new_tag_ids = Tag.where(name: new_value.split("|")).pluck(:id)
+    now = Time.zone.now
+
+    notification_level = tag_notification_level(id)
+
+    TagUser
+      .where(tag_id: (previous_tag_ids - new_tag_ids), notification_level: notification_level)
+      .in_batches(of: batch_size) { |batch| batch.delete_all }
+
+    modified_tags = new_tag_ids - previous_tag_ids
+
+    skip_user_ids = {}
+    users_scope = {}
+
+    modified_tags.each do |tag_id|
+      skip_user_ids[:tag_id] = TagUser.where(tag_id: tag_id).pluck(:user_id)
+      users_scope[:tag_id] = User.real.where(staged: false).where.not(id: skip_user_ids[:tag_id])
+    end
+
+    total_users_to_process = users_scope.values.map(&:count).sum
+    processed_total = 0
+
+    modified_tags.each do |tag_id|
+      skip_user_ids[:tag_id] = TagUser.where(tag_id: tag_id).pluck(:user_id)
+      users_scope[:tag_id] = User.real.where(staged: false).where.not(id: skip_user_ids[:tag_id])
+
+      users_scope[:tag_id]
+        .select(:id)
+        .find_in_batches(batch_size: batch_size) do |users|
+          tag_users =
+            users.map do |user|
+              {
+                tag_id: tag_id,
+                user_id: user.id,
+                notification_level: notification_level,
+                created_at: now,
+                updated_at: now,
+              }
+            end
+
+          TagUser.insert_all!(tag_users)
+
+          processed_total += users.size
+          publish_progress(id, processed_total, total_users_to_process, modified_tags)
+        end
+    end
+
+    publish_progress(id, processed_total, total_users_to_process, modified_tags)
+  end
+
+  def self.publish_progress(
+    site_setting_name,
+    processed_total = 0,
+    total_users_to_process = 0,
+    modified = nil
+  )
+    status =
+      if modified.empty? || processed_total >= total_users_to_process
+        "completed"
+      else
+        "enqueued"
+      end
+
+    progress = modified.empty? ? nil : "#{processed_total}/#{total_users_to_process}"
+
+    MessageBus.publish(
+      "/site_setting/#{site_setting_name}/process",
+      status: status,
+      progress: progress,
+      group_ids: [Group::AUTO_GROUPS[:admin]],
+    )
+  end
```

Comment:
This is a new change that we are adding so we should update our test coverage to reflect this change.

Reply:
Yes I stashed this when doing something else and forgot about it completely

Reply:
Hmm I still don't see the tests for this change. Since we are publishing a message bus message, we can ensure that we are publishing the right message and to the right group by using the `MessageBus.track_publish` test helper. You can refer to an example like the following:

https://github.com/discourse/discourse/blob/e6f9cde35e3f68d4ffb0a46a8361ce099a558ac6/spec/lib/presence_channel_spec.rb#L92-L100

Reply:
@jdmartinez1062 It doesn't seem like we are testing for MessageBus changes yet.